pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java

```diff
@@ -152,39 +202,82 @@ private TransferableBlock buildJoinedDataBlock(TransferableBlock leftBlock)
       return _upstreamErrorBlock;
-    } else if (TransferableBlockUtils.isNoOpBlock(leftBlock) || TransferableBlockUtils.isEndOfStream(leftBlock)) {
+    }
+    if (leftBlock.isNoOpBlock() || (leftBlock.isSuccessfulEndOfStreamBlock() && !needUnmatchedRows())) {
       return leftBlock;
     }
+    if (leftBlock.isSuccessfulEndOfStreamBlock()) {
+      // Return remaining non-matched rows for non-inner join.
+      List<Object[]> returnRows = new ArrayList<>();
+      if (needUnmatchedRightRows()) {
+        for (Map.Entry<Key, List<Object[]>> entry : _broadcastRightTable.entrySet()) {
+          Set<Integer> matchedIdx = _matchedRightRows.getOrDefault(entry.getKey(), new HashSet<>());
+          int rowSize = entry.getValue().size();
+          Set<Integer> allIdx = IntStream.range(0, rowSize).boxed().collect(Collectors.toSet());
+          Set<Integer> unmatchedIdx = Sets.difference(allIdx, matchedIdx);
+          for (int unmatchedRow : unmatchedIdx) {
+            Object[] resultRow = joinRow(null, entry.getValue().get(unmatchedRow));
+            returnRows.add(resultRow);
+          }
+        }
+      }
+      if (needUnmatchedLeftRows()) {
+        for (Map.Entry<Key, List<Object[]>> entry : _unmatchedLeftRows.entrySet()) {
+          for (Object[] unmatchedLeftRow : entry.getValue()) {
+            Object[] resultRow = joinRow(unmatchedLeftRow, null);
+            returnRows.add(resultRow);
+          }
+        }
+      }
+      _isTerminated = true;
+      return new TransferableBlock(returnRows, _resultSchema, DataBlock.Type.ROW);
+    }
     List<Object[]> rows = new ArrayList<>();
     List<Object[]> container = leftBlock.isEndOfStreamBlock() ? new ArrayList<>() : leftBlock.getContainer();
     for (Object[] leftRow : container) {
+      Key key = new Key(_leftKeySelector.getKey(leftRow));
       // NOTE: Empty key selector will always give same hash code.
-      List<Object[]> hashCollection =
-          _broadcastHashTable.getOrDefault(new Key(_leftKeySelector.getKey(leftRow)), Collections.emptyList());
-      // If it is a left join and right table is empty, we return left rows.
-      if (hashCollection.isEmpty() && _joinType == JoinRelType.LEFT) {
-        rows.add(joinRow(leftRow, null));
-      } else {
-        // If it is other type of join.
-        for (Object[] rightRow : hashCollection) {
-          // TODO: Optimize this to avoid unnecessary object copy.
-          Object[] resultRow = joinRow(leftRow, rightRow);
-          if (_joinClauseEvaluators.isEmpty() || _joinClauseEvaluators.stream().allMatch(evaluator ->
-              (Boolean) FunctionInvokeUtils.convert(evaluator.apply(resultRow), DataSchema.ColumnDataType.BOOLEAN))) {
-            rows.add(resultRow);
+      List<Object[]> matchedRightRows = _broadcastRightTable.getOrDefault(key, null);
+      if (matchedRightRows == null) {
+        if (needUnmatchedLeftRows()) {
+          rows.add(joinRow(leftRow, null));
+        }
```

Comment:
why do we do this here. and then again in line 224-228?

Reply:
Actually, it is an issue caused by my previous implementation. I updated this

Reply:
Updated. PTAL